sklearn/tree/_tree.pyx

```diff
@@ -1132,3 +1132,284 @@ cdef class Tree:
+                n_leaves[parent_idx] += 1
+                leaf_idx = parent_idx
+
+    cdef:
+        # candidate nodes that can be pruned
+        bint[:] candidate_nodes = view.array(
+            shape=(n_nodes, ), itemsize=sizeof(bint), format="i")
+        # nodes in subtree
+        bint[:] in_subtree = view.array(
+            shape=(n_nodes, ), itemsize=sizeof(bint), format="i")
+        DOUBLE_t[:] g_node = np.zeros(shape=n_nodes, dtype=np.float64)
+        SIZE_t cur_idx
+        DOUBLE_t subtree_alpha
+        DOUBLE_t cur_ccp_alpha
+        SIZE_t child_l_idx
+        SIZE_t child_r_idx
+        SIZE_t n_pruned_leaves
+        DOUBLE_t r_diff
+        DOUBLE_t max_float64 = np.finfo(np.float64).max
+        SIZE_t init_capacity = 0
+
+    with nogil:
+        for i in range(leaves_in_subtree.shape[0]):
+            candidate_nodes[i] = not leaves_in_subtree[i]
+
+        for i in range(n_nodes):
+            in_subtree[i] = 1
+
+        while True:
+            # If root node is a leaf
+            if not candidate_nodes[0]:
+                break
+
+            # computes ccp_alpha for subtrees and finds the minimal alpha
+            cur_ccp_alpha = max_float64
+            for i in range(n_nodes):
+                if not candidate_nodes[i]:
+                    continue
+                subtree_alpha = (r_node[i] - r_branch[i]) / (n_leaves[i] - 1)
+                if subtree_alpha < cur_ccp_alpha:
+                    cur_ccp_alpha = subtree_alpha
+                    cur_idx = i
+
+            # The subtree on the previous iteration has the greatest ccp_alpha
+            # less than or equal to self.ccp_alpha
+            if cur_ccp_alpha > ccp_alpha:
+                break
+
+            # stack uses only the start variable
+            rc = stack.push(cur_idx, 0, 0, 0, 0, 0, 0)
+            if rc == -1:
+                with gil:
+                    raise MemoryError()
+
+            # descendants of branch are not in subtree
+            while not stack.is_empty():
+                stack.pop(&stack_record)
+                node_idx = stack_record.start
+                candidate_nodes[node_idx] = 0
+                leaves_in_subtree[node_idx] = 0
+                in_subtree[node_idx] = 0
+
+                child_l_idx = child_l[node_idx]
+                child_r_idx = child_r[node_idx]
+                if child_l_idx != child_r_idx:
+                    stack.push(child_l_idx, 0, 0, 0, 0, 0, 0)
+                    stack.push(child_r_idx, 0, 0, 0, 0, 0, 0)
+            leaves_in_subtree[cur_idx] = 1
+            in_subtree[cur_idx] = 1
+
+            # updates number of leaves
+            n_pruned_leaves = n_leaves[cur_idx] - 1
+            n_leaves[cur_idx] = 0
+
+            # # computes the increase in r_branch to bubble up
+            r_diff = r_node[cur_idx] - r_branch[cur_idx]
+            r_branch[cur_idx] = r_node[cur_idx]
+
+            # bubble up values to ancestors
+            cur_idx = parent[cur_idx]
```

Comment:
nitpicking: up until here, I would rename `cur_idx` into `pruned_branch_node_idx` (or something of the like that'd be more explicit), and from here to the end rename `cur_idx` into `node_idx` since `node_idx` is used as a loop variable above.

Or I guess I'd just think it'd be nice to make clear that `cur_idx` is the index of the node whose branch is going to be pruned.